### PR TITLE
osdeps: put OS dependencies information in repository

### PR DESCRIPTION
**Pull request by hakamine**
Add a directory osdeps/ for each archivematica component
(archivematicaCommon, dashboard, MCPServer, MCPClient) with files
containing information of OS dependencies. To be used in deployment
(ansible, etc.)
( Currently, files in osdeps/ directory are used by [PR104 of the ansible role](https://github.com/artefactual-labs/ansible-archivematica-src/pull/104) to be able to deploy qa/1.6.x on CentOS/RH 7 in addition to Ubuntu trusty )

### CODE REVIEW

File: src/MCPClient/osdeps/Ubuntu-14.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    "deb [arch=amd64] http://packages.archivematica.org/1.6.x/ubuntu-externals trusty main",
+    "deb http://archive.ubuntu.com/ubuntu/ trusty multiverse",
+    "deb http://archive.ubuntu.com/ubuntu/ trusty-security universe",
+    "deb http://archive.ubuntu.com/ubuntu/ trusty-updates multiverse"
+  ],
+  "osdeps_packages": [
+  { "name": "atool", "state": "latest"},
+  { "name": "bagit", "state": "latest"},
+  { "name": "bulk-extractor", "state": "latest"},
+  { "name": "ffmpeg", "state": "latest"},
```
Comment (Hwesta) [inline src/MCPClient/osdeps/Ubuntu-14.json:22]: Is there a way to specify minimum version? The control file lists `>=7:30`

File: src/MCPClient/osdeps/RedHat-7.json [hunk 2]
```diff
@@ -0,0 +1,65 @@
+{
```
Comment (Hwesta) [inline src/MCPClient/osdeps/RedHat-7.json]: The RedHat and CentOS json lists are identical - do we need both?

File: src/MCPServer/osdeps/Ubuntu-14.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    "Include here keys, repos, and packages required by this component not already in archivematicaCommon osdeps file"
+  ],
+  "osdeps_repokeys": [
+  ],
+  "osdeps_repos": [
+  ],
+  "osdeps_packages": [
+  { "name": "dbconfig-common", "state": "latest"},
+  { "name": "logapp", "state": "latest"},
+  { "name": "uuid", "state": "latest"}
```
Comment (Hwesta) [inline src/MCPServer/osdeps/Ubuntu-14.json:13]: Should this also install gearman & a database?

File: src/dashboard/osdeps/Ubuntu-14.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  "_comments": [
+    "Add any comments here",
+    "Include here keys, repos, and packages required by this component not already in archivematicaCommon osdeps file"
+  ],
+  "osdeps_repokeys": [
+  ],
+  "osdeps_repos": [
+  ],
+  "osdeps_packages": [
+  { "name": "gettext", "state": "latest"}
```
Comment (Hwesta) [inline src/dashboard/osdeps/Ubuntu-14.json:11]: There are several dependencies listed in the control file that are not in here - are they handled elsewhere?

**Review by Hwesta — COMMENTED** [review]
Are we planning to replace the debian control files with this, or is this intended to be just for Ansible?  How do we plan to keep the dependencies & versions in sync?

File: src/MCPServer/osdeps/Ubuntu-14.json (hunk 3 above)
Comment (scollazo) [reply src/MCPServer/osdeps/Ubuntu-14.json:13]: I don't think so. Gearman, elasticsearch and mysql are services archivematica relies on, but they should be handled on their own. 

I'm thinking about environments with their own mysql/es servers, were installing the db servers as part of archivematica doesn't make sense.

File: src/MCPClient/osdeps/Ubuntu-14.json (hunk 1 above)
Comment (scollazo) [reply src/MCPClient/osdeps/Ubuntu-14.json:22]: I also think that we should add a known working version as minimum, keeping things explicit helps in the long term.

**Comment by scollazo** [comment]
@Hwesta , yes, we plan to populate the `Depends:` field in debian/control files with the content provided by those. This should help handle different dependency versions for ubuntu 14.04 and 16.04, without having different control files for each distribution.

File: src/MCPServer/osdeps/Ubuntu-14.json (hunk 3 above)
Comment (Hwesta) [reply src/MCPServer/osdeps/Ubuntu-14.json:13]: That makes sense - they don't need to be installed on the same machine, just available.

**Comment by hakamine** [comment]
>Are we planning to replace the debian control files with this, or is this intended to be just for Ansible? How do we plan to keep the dependencies & versions in sync?

@Hwesta The idea is to try to put the dependencies in one place and reuse as much as possible (not only by ansible). That said, the current format of the .json files are made so that they are easily usable by ansible (I think those could probably be readily usable if we try to make containers with them).

**Comment by Hwesta** [comment]
Ah, great. Sound like you have that covered.

I noticed the RedHat & CentOS lists are identical - do they both need to exist?

File: src/MCPClient/osdeps/Ubuntu-14.json (hunk 1 above)
Comment (hakamine) [reply src/MCPClient/osdeps/Ubuntu-14.json:22]: The current format for the .json file is based on what the ansible apt / yum modules takes as input (which in turn is based on the way apt and yum OS commands work), it is possible to specify a fixed version or to install the "latest" version available from the repository, but not possible to specify a "minimum" version. 
As a workaround I think that we could add some parameters to the .json file to be used by the package building scripts (example `{ "name": "ffmpeg", "state": "latest", "debcontrol":">= 7:3.0"}`)

File: src/MCPClient/osdeps/RedHat-7.json (hunk 2 above)
Comment (hakamine) [reply src/MCPClient/osdeps/RedHat-7.json]: I expected there could be some small differences between RH and CentOS but so far this wasn't the case. We could probably use a symlink to one for now (and split into separate files if we find any difference in the future)

File: src/dashboard/osdeps/Ubuntu-14.json (hunk 4 above)
Comment (hakamine) [reply src/dashboard/osdeps/Ubuntu-14.json:11]: I think some of the dependencies listed in the current control files may not be required. I tested deploying archivematica from source with the dependencies in the osdeps files in this PR and so far haven't found issues.

**Comment by hakamine** [comment]
> I noticed the RedHat & CentOS lists are identical - do they both need to exist?

Currently the ansible role uses the variable `ansible_distribution` ( which gets the OS the target system is using ) to pick the correct osdeps .json file to use for deployment.
I expected there could be some small differences between RH and CentOS but so far this wasn't the case. We could probably use a symlink to one of them instead of having two files with duplicate content (and remove the symlink and have two separate files if we find any difference in the future).

File: src/dashboard/osdeps/Ubuntu-14.json (hunk 4 above)
Comment (Hwesta) [reply src/dashboard/osdeps/Ubuntu-14.json:11]: Good to know. Maybe worthwhile to consider removing them from the control file after this release?

**Comment by Hwesta** [comment]
Ahh, Ansible wants the two versions.  I like the symlink idea.

**Review by Hwesta — APPROVED** [review]

**Comment by hakamine** [comment]
Merged to qa/1.6.x and also cherry-picked to stable/1.6.x

**Comment by hakamine** [comment]
Also cherry picked to qa/1.x

**Pull request merged**